cmds/dhclient/dhclient.go

```diff
@@ -0,0 +1,191 @@
+	// This is truly amazing but /sys appends newlines to all this data.
+	n = bytes.TrimSpace(n)
+	m, err := net.ParseMAC(string(n))
+	if err != nil {
+		done <- fmt.Errorf("mac error: %v", err)
+		return
+	}
+
+	iface, err := netlink.LinkByName(ifname)
+	if err != nil {
+		done <- fmt.Errorf("%s: netlink.LinkByName failed: %v", ifname, err)
+		return
+	}
+
+	c, err := dhcp4client.NewInetSock(dhcp4client.SetLocalAddr(net.UDPAddr{IP: net.IPv4(0, 0, 0, 0), Port: 68}), dhcp4client.SetRemoteAddr(net.UDPAddr{IP: net.IPv4bcast, Port: 67}))
+	if err != nil {
+		done <- fmt.Errorf("client conection generation: %v", err)
+		return
+	}
+
+	client, err := dhcp4client.New(dhcp4client.HardwareAddr(m), dhcp4client.Connection(c), dhcp4client.Timeout(timeout))
+	if err != nil {
+		done <- fmt.Errorf("error: %v", err)
+		return
+	}
+
+	// We require at least one successful request.
+	success, packet, err := client.Request()
+
+	if err != nil {
+		networkError, ok := err.(*net.OpError)
+		if ok && networkError.Timeout() {
+			done <- fmt.Errorf("%s: did not find a DHCP Server", n)
+			return
+		}
+		done <- fmt.Errorf("%s: Error:%v", n, err)
+		return
+	}
+
+	debug("Success on %s: %v\n", n, success)
+	debug("Packet: %v\n", packet)
+	debug("Lease is %v seconds\n", packet.Secs())
+
+	if !success {
+		done <- fmt.Errorf("we did not sucessfully get a DHCP Lease?")
+		return
+	}
+	log.Printf("IP Received: %v\n", packet.YIAddr().String())
+
+	for i := 0; i < *renewals; i++ {
+		addr := packet.YIAddr()
+		// We got here because we got a good packet.
+		o := packet.ParseOptions()
+		netmask, ok := o[dhcp4.OptionSubnetMask]
+		if ok {
+			log.Printf("OptionSubnetMask is %v\n", netmask)
+		} else {
+			// what do to?
+			netmask = addr
+		}
+		dst := &netlink.Addr{IPNet: &net.IPNet{IP: packet.YIAddr(), Mask: netmask}, Label: ""}
+		// Add the address to the iface.
+		if err := netlink.AddrAdd(iface, dst); err != nil {
+			if os.IsExist(err) {
+				done <- fmt.Errorf("add %v to %v: %v", dst, n, err)
+				return
+			}
+		}
+
+		if gwData, ok := o[dhcp4.OptionRouter]; ok {
+			log.Printf("router %v", gwData)
+			routerName := net.IP(gwData).String()
+			debug("routerName %v", routerName)
+			r := &netlink.Route{
+				Dst:       &net.IPNet{IP: packet.GIAddr(), Mask: netmask},
+				LinkIndex: iface.Attrs().Index,
+				Gw:        packet.GIAddr(),
+			}
+
+			if err := netlink.RouteReplace(r); err != nil {
+				done <- fmt.Errorf("%s: add %s: %v", ifname, r.String(), routerName)
+				return
+			}
+		}
+
+		// We can not assume the server will give us any grace time.
+		// So sleep for just a tiny bit less than the minimum.
+		time.Sleep(timeout - slop)
+		debug("Start Renewing Lease")
+		success, packet, err = client.Renew(packet)
+		if err != nil {
+			networkError, ok := err.(*net.OpError)
+			if ok && networkError.Timeout() {
+				done <- fmt.Errorf("Renewal Failed! Because it didn't find the DHCP server very Strange")
+				return
+			}
+			done <- fmt.Errorf("error: %v", err)
```

Comment:
return here

on that topic, to avoid this kind of mistake, this function should really be

```go
func dhclient(...) error {
   return fmt.Errorf("..")
}
```

and below should be

```go
go func(done chan error) {
  done <- dhclient(s, ...)
}
```